torch/fx/experimental/proxy_tensor.py

```diff
@@ -74,6 +74,70 @@ def decompose(decomposition_table):
 no_default = object()
 
 py_sym_types = (SymInt, SymFloat, SymBool)
+
+
+@dataclass
+class _SymExprHash:
+    """
+    Hash for a py_sym_types that will use the underlying sympy expression
+    """
+
+    sym_obj: py_sym_types
+
+    def __hash__(self) -> int:
+        return hash((type(self.sym_obj), self.sym_obj.node.expr))
+
+    def __eq__(self, value: "_SymExprHash") -> bool:
+        assert isinstance(value, _SymExprHash)
+        return self.sym_obj.node.expr == value.sym_obj.node.expr
+
+
+class _SymHashingDict:
```

Comment:
This implementation seems way stronger than what your test is exercising. What you are doing here is something like if I compute s0 + s1 multiple times, I will discover and CSE all of these into a single expression. The test, OTOH, seems to imply that if you did the CSE only on expressions that were merely Symbol, that would have been sufficient. Do you actually care about more complex CSE? Because if you don't, I think this code gets a lot simpler.

Reply:
I think the more complex CSE here is good.. I'm not sure how much simpler it gets if I don't attempt to it. I could not rely on the Sympy hash implementation but otherwise should be pretty similar ?

Reply:
I can add more tests